uc-crux-llvm/src/UCCrux/LLVM/Run/Unsoundness.hs

```diff
@@ -39,39 +46,60 @@ data WithUnsoundness a = WithUnsoundness
 
 data Unsoundness = Unsoundness
   { unsoundOverridesUsed :: Set UnsoundOverrideName,
-    unsoundSkipOverridesUsed :: Set SkipOverrideName
+    unsoundSkipOverridesUsed :: Set SkipOverrideName,
+    unsoundSpecsUsed :: Set SpecUse
   }
   deriving (Eq, Ord, Show)
 
-ppUnsoundness :: Unsoundness -> Doc Void
-ppUnsoundness u =
+ppUnsoundness ::
+  -- | What kind of unsoundness do you care about? For instance, if this is
+  -- 'Soundness.Overapprox', this function will print only things that are
+  -- unsound for verification.
+  Soundness ->
+  Unsoundness ->
+  Doc Void
+ppUnsoundness s u =
   PP.nest 2 $
     PP.vcat $
-      ( PP.pretty
-          "The following unsound overrides (built-in functions) were used:" :
-        bullets
-          (map getUnsoundOverrideName (Set.toList (unsoundOverridesUsed u)))
-      )
-        ++ ( PP.pretty
-               "Execution of the following functions was skipped:" :
-             bullets
-               (map getSkipOverrideName (Set.toList (unsoundSkipOverridesUsed u)))
-           )
+      concat
+        [ PP.pretty
+            "The following unsound overrides (built-in functions) were used:" :
+              bullets
+                (map getUnsoundOverrideName (Set.toList (unsoundOverridesUsed u)))
+        , PP.pretty
+            "Execution of the following functions was skipped:" :
+              bullets
+                (map getSkipOverrideName (Set.toList (unsoundSkipOverridesUsed u)))
+        , PP.pretty
+            "The following unsound specifications were applied:" :
+              bullets specs
+        ]
   where
     bullets = map ((PP.pretty "-" PP.<+>) . PP.pretty)
+    specs =
+      mapMaybe
+        (\(OvSpec.SpecUse nm s') ->
+           if Sound.atLeastAsSound s' s
+           then Nothing
+           else Just (Text.pack (functionNameToString nm)
+                        <> Text.pack ": "
+                        <> Text.pack (Sound.soundnessToString s')))
+        (Set.toList (unsoundSpecsUsed u))
 
 instance Semigroup Unsoundness where
   u1 <> u2 =
     Unsoundness
       { unsoundOverridesUsed =
           unsoundOverridesUsed u1 <> unsoundOverridesUsed u2,
         unsoundSkipOverridesUsed =
-          unsoundSkipOverridesUsed u1 <> unsoundSkipOverridesUsed u2
+          unsoundSkipOverridesUsed u1 <> unsoundSkipOverridesUsed u2,
+        unsoundSpecsUsed = unsoundSpecsUsed u1 <> unsoundSpecsUsed u2
       }
 
 instance Monoid Unsoundness where
   mempty =
     Unsoundness
       { unsoundOverridesUsed = Set.empty,
-        unsoundSkipOverridesUsed = Set.empty
+        unsoundSkipOverridesUsed = Set.empty,
```

Comment:
Minor suggestion: using `mempty` avoids module prefixing and is robust against changing the underlying container type.  Occurs in lots of other places if you decide to make this change.

Reply:
> avoids module prefixing

I don't see this as a negative thing, is this an aesthetic preference?

> is robust against changing the underlying container

Agreed, though changing to `mempty` countervails [my preference to make things minimally polymorphic](https://langston-barrett.github.io/style/haskell.html) for the sake of explicitness and readability. I think `mempty` vs. `Set.empty` is pretty low-stakes, but I'll probably just keep it as is for now.